### PR TITLE
Always clean up FailedToScheduleReplica with wrong HardNodeAffinity

### PR DESCRIPTION
**Pull request by ejweber**
#### Which issue(s) this PR fixes:

longhorn/longhorn#8522

#### What this PR does / why we need it:

Delete a failed to schedule replica with HardNodeAffinity if DataLocality is disabled.

Previously, we would only do this if there were enough healthy replicas, but it led to the "deadlock" in longhorn/longhorn#8522 where we would not schedule more replicas until we deleted the failed to schedule one, but we would not delete the failed to schedule one until there were enough healthy ones.

#### Special notes for your reviewer:

I experimented with changing the behavior of this function even more, but there were some weird side effects. I think it is better to keep it working almost exactly as it did before. I did, however, rearrange a bit.

### CODE REVIEW

**Comment by ejweber** [comment]
## Test steps

Follow the reproduce steps in https://github.com/longhorn/longhorn/issues/8522#issuecomment-2105011487.

6. Observe that the failed to schedule replica is deleted and a new one (that can be scheduled) replaces it.

**Comment by ejweber** [comment]
Regression testing in: https://ci.longhorn.io/job/private/job/longhorn-tests-regression/6924/.

File: controller/volume_controller.go [hunk 1]
```diff
@@ -1028,24 +1028,38 @@ func (c *VolumeController) cleanupCorruptedOrStaleReplicas(v *longhorn.Volume, r
 	return nil
 }
 
-func (c *VolumeController) cleanupFailedToScheduledReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (err error) {
+func (c *VolumeController) cleanupFailedToScheduleReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (err error) {
```
Comment (james-munson) [inline controller/volume_controller.go:1031]: While you're at it, the name could change to cleanupFailedToScheduleReplica, since it used to do multiple replicas, and now it does the first such it finds.

Comment (james-munson) [reply controller/volume_controller.go:1031]: (Personally, I found the name hard to parse, until I mentally translated it as cleanupUnscheduledReplica.)

**Review by james-munson — COMMENTED** [review]
One question.

File: controller/volume_controller.go (hunk 1 above)
Comment (ejweber) [reply controller/volume_controller.go:1031]: TBH, this is a bit of an oversight on my part. When I read the original code, I thought it deleted one replica at a time as well. Reading it back again, I'm not sure why. I'm inclined to return to the original behavior and allow multiple deletions in a single reconcile. WDYT?

Comment (james-munson) [reply controller/volume_controller.go:1031]: I agree, I would stick with the original behavior unless there is a pressing reason to change, and I don't see one.

File: controller/volume_controller.go [hunk 2]
```diff
@@ -1028,24 +1028,38 @@ func (c *VolumeController) cleanupCorruptedOrStaleReplicas(v *longhorn.Volume, r
...
+				replicaToCleanUp = r
+				break
 			}
 		}
 	}
 
+	if replicaToCleanUp != nil {
+		logrus.Infof("Cleaning up failed to scheduled replica %v", replicaToCleanUp.Name)
+		if err := c.deleteReplica(replicaToCleanUp, rs); err != nil {
+			return errors.Wrapf(err, "failed to cleanup failed to scheduled replica %v", replicaToCleanUp.Name)
```
Comment (james-munson) [inline controller/volume_controller.go:1059]: Really tiny nit, but could you make this error text "failed to cleanup failed-to-schedule replica %v".

File: controller/volume_controller.go (hunk 1 above)
Comment (james-munson) [reply controller/volume_controller.go:1031]: Looks good.

**Review by james-munson — APPROVED** [review]
LGTM.

**Review by shuo-wu — APPROVED** [review]

**Comment by ejweber** [comment]
@mergify backport v1.5.x

**Comment by ejweber** [comment]
@mergify backport v1.6.x
